mountpoint-s3/src/data_cache/disk_data_cache.rs

```diff
@@ -0,0 +1,312 @@
+//! Module for the on-disk data cache implementation.
+
+use std::fs;
+use std::io::{ErrorKind, Write};
+use std::ops::RangeBounds;
+use std::path::PathBuf;
+
+use base64ct::{Base64Url, Encoding};
+use bytes::{BufMut, BytesMut};
+use mountpoint_s3_crt::checksums::crc32c;
+use tracing::{error, trace, warn};
+
+use crate::data_cache::DataCacheError;
+
+use super::{BlockIndex, CacheKey, ChecksummedBytes, DataCache, DataCacheResult};
+
+/// On-disk implementation of [DataCache].
+///
+/// TODO: Store checksums on disk, reconstruct as [ChecksummedBytes] using same checksum avoiding recomputation.
+///
+/// TODO: Store additional metadata with each block such as expected S3 key, ETag, etc..
+pub struct DiskDataCache {
+    block_size: u64,
+    cache_directory: PathBuf,
+}
+
+impl DiskDataCache {
+    /// Create a new instance of an [DiskDataCache] with the specified `block_size`.
+    pub fn new(cache_directory: PathBuf, block_size: u64) -> Self {
+        DiskDataCache {
+            block_size,
+            cache_directory,
+        }
+    }
+
+    /// Get the relative path for the given block.
+    fn get_path_for_key(&self, cache_key: &CacheKey) -> PathBuf {
+        let mut path = self.cache_directory.join("v1");
+
+        // An S3 key may be up to 1024 UTF-8 bytes long, which exceeds the maximum UNIX file name length.
+        // Instead, we encode the key and split into 255 character long directory names.
+        let encoded_s3_key = Base64Url::encode_string(cache_key.s3_key.as_bytes());
+        let mut slice = encoded_s3_key.as_str();
+        while !slice.is_empty() {
+            let (chunk, remaining) = slice.split_at(255.min(slice.len()));
+            path.push(chunk);
+            slice = remaining;
+        }
```

Comment:
I am reconsidering if we actually need the base 64 encoding. It does obfuscate the path a bit, but I'm not convinced we actually have any other reason at this point.

Instead, could we just put the blocks in directories based on the actually file path they represent? (Of course, that makes this a cache suited towards files with etag rather than objects with an etag.)

Reply:
I like the idea, but it may be a bit cumbersome to pass the file path here (or the mounted prefix?). Happy to leave the base64 encoding for now.

Reply:
Obfuscation is important here — we don't want customers looking inside this directory instead of going through Mountpoint. But it seems like we could just take a hash instead. It's also often a good idea to split the directory into subdirectories (based on the first few characters of the hash) to avoid hitting any filesystem-specific maximum number of directory entries.

Reply:
Updating to hash now. I'll take splitting directories into subdirs as a TODO.